### PR TITLE
Fix exception during textInfos.moveToCodepointOffset when some characters correspond to 0 pythonic characters

### LINKED ISSUE #16406: Navigating to different style text in MS outlook, NVDA often cannot find desired offset in TextInfo.
### Steps to reproduce:
1. Open NVDA
2. Assign [ and shift+[ for different style navigation 
3. Assign ] and shift+] for same style navigation
4. Open an Outlook email, see email attached, just remove the .zip characters at the end to open the email.
5. Make sure UIA for MS Word is disabled, this has also an effect on MS Outlook.
6. Press the assigned commands for navigating to next and previous text styles

### Actual behavior:
NVDA often gives a runtime error when attempting to navigate to next or previous different style text and is silent.
```
IO - inputCore.InputManager.executeGesture (22:50:08.993) - winInputHook (10508):
Input: kb(laptop):ü
ERROR - scriptHandler.executeScript (22:50:09.164) - MainThread (19244):
error executing script: <bound method BrowseModeTreeInterceptor.addQuickNav.<locals>.<lambda> of <appModules.outlook.MailViewerTreeInterceptor object at 0x0A1C4630>> with gesture 'ü'
Traceback (most recent call last):
  File "scriptHandler.pyc", line 295, in executeScript
  File "browseMode.pyc", line 558, in <lambda>
  File "browseMode.pyc", line 516, in _quickNavScript
  File "browseMode.pyc", line 2447, in _iterTextStyle
  File "textInfos\__init__.pyc", line 793, in moveToCodepointOffset
RuntimeError: Unable to find desired offset in TextInfo.
```
[Webinar des Monats Förderung durch die öffentliche Hand.msg.zip](https://github.com/nvaccess/nvda/files/15002371/Webinar.des.Monats.Forderung.durch.die.offentliche.Hand.msg.zip)

### Expected behavior:
NVDA should report the next or the previous different style text properly,  or "no next different style text" and "no previous different style text", in case no such text style has been found. This works as expected for next and previous same style text commands.

### NVDA logs, crash dumps and other attachments:
n/a

#### NVDA installed/portable/running from source:
Installed

#### NVDA version:
Version: alpha-31612,3d4aOa8f (2024.2.0.31612)

#### Windows version:
Windows 11 23 H2

#### Name and version of other software in use when reproducing the issue:
Microsoft 365 MSO (Version 2403 Build 16.0.17425.20176) 64 Bit

#### Other information about your system:
n/a

#### Does the issue still occur after restarting your computer?
yes

#### Have you tried any other versions of NVDA? If so, please report their behaviors.
n/a

#### If NVDA add-ons are disabled, is your problem still occurring?
yes

#### Does the issue still occur after you run the COM Registration Fixing Tool in NVDA's tools menu?
yes

### PR DESCRIPTION
**Pull request by mltony**
### Link to issue number:
Partially closes #16406.
### Summary of the issue:
When trying to navigate by style an exception is thrown:
```
RuntimeError: Unable to find desired offset in TextInfo.
```
### Description of user facing changes
N/A
### Description of development approach
Investigation showed that in Outlook with word UIA enabled, sometimes there are characters that translate into 0 codepoint characters. To illustrate this further:
```
>>> p=focus.treeInterceptor.makeTextInfo('caret')
>>> p.expand('paragraph')
>>> p.text
'View in browser'
>>> p.collapse(end=True)
>>> p.move('character', -1, 'start')
-1
>>> p.text
''
>>> len(p.text)
0
```

E.g. we move a collapsed text info start by -1 character, and the resulting textInfo still appears to be empty- at least its Python text is empty.
I anticipated this to happen somewhere, but didn't encounter a single example when I was implementing `moveToCodepointOffset` function, so I missed one edge case. Now fixing this edge case. Also adding unit tests.
### Testing strategy:
Unit test.
Tested on test case from the original issue with UIA enabled.
### Known issues with pull request:
N/A
### Code Review Checklist:

- [x] Documentation:
  - Change log entry
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] API is compatible with existing add-ons.
- [x] Security precautions taken.

### CODE REVIEW

File: source/textInfos/__init__.py [hunk 1]
```diff
@@ -825,10 +827,15 @@ def moveToCodepointOffset(
...
-			if actualCodepointOffset <= 0 or actualCodepointOffset >= totalCodepointOffset:
+			if (
+				(not moveFromLeft and actualCodepointOffset <= 0)
+				or (moveFromLeft and actualCodepointOffset >= totalCodepointOffset)
+			):
 				# We overshot, call this recursion attempt failed and try again lower movement
 				lastRecursed = 0
 				continue
+			if actualCodepointOffset < 0 or actualCodepointOffset > totalCodepointOffset:
+				raise RuntimeError(f"{actualCodepointOffset =} went out of the boundaries; {totalCodepointOffset=}")
```
Comment (seanbudd) [inline source/textInfos/__init__.py:838]:
```suggestion
				raise RuntimeError(f"{actualCodepointOffset=} went out of the boundaries; {totalCodepointOffset=}")
```

**Review by seanbudd — DISMISSED** [review]

**Comment by LeonarddeR** [comment]
@mltony Could you please look into the following issue, which seems related? Still reproducible wit this pr?

In word, type: `1. Hello`

Then

```
>>> info = focus.makeTextInfo("all")
>>> info.text
'1. hello'
>>> info.moveToCodepointOffset(0)
<NVDAObjects.UIA.wordDocument.WordDocumentTextInfo object at 0x0AF75050>
>>> info.moveToCodepointOffset(1)
Traceback (most recent call last):
  File "<console>", line 1, in <module>
  File "D:\Development\A11y\nvda\source\textInfos\__init__.py", line 794, in moveToCodepointOffset
    raise RuntimeError("Unable to find desired offset in TextInfo.")
RuntimeError: Unable to find desired offset in TextInfo.
>>> info.moveToCodepointOffset(2)
Traceback (most recent call last):
  File "<console>", line 1, in <module>
  File "D:\Development\A11y\nvda\source\textInfos\__init__.py", line 794, in moveToCodepointOffset
    raise RuntimeError("Unable to find desired offset in TextInfo.")
RuntimeError: Unable to find desired offset in TextInfo.
>>> info.moveToCodepointOffset(3)
Traceback (most recent call last):
  File "<console>", line 1, in <module>
  File "D:\Development\A11y\nvda\source\textInfos\__init__.py", line 818, in moveToCodepointOffset
    raise RuntimeError("Unable to find desired offset in TextInfo.")
RuntimeError: Unable to find desired offset in TextInfo.
>>> info.moveToCodepointOffset(4)
<NVDAObjects.UIA.wordDocument.WordDocumentTextInfo object at 0x06DE32F0>
```

Is this expected behavior, or should the function for values 1, 2 and 3 just return the collapsed start of the textInfo?

**Comment by mltony** [comment]
@LeonarddeR,
That's expected. That's because in MSWord in a list the first character (when moving by `textInfos.UNIT_CHARACTER`) corresponds to 4 charactters in Python: "1. H". Thus binary search fails as expected.
I would strongly argue against quietly swallowing that exception and trying to resolve it somehow inside `moveToCodepointOffset` function. For example if we do it, then during sentence navigation we might get stuck at a particular sentence when pressing `alt+DownArrow` because `moveToCodepointOffset` would quietly take us a few characters back every time. Dealing with this problem is hard but it needs to be dealt with outside `moveToCodepointOffset` depending on your use case.
One thing I regret though is raising generic RuntimeError - I should've declared a custom exception class so that it's easy to catch.
